Move goprotoc logic to app/goprotoc and refactor all control flow to use errors and not os.Exit

bufdev opened the pull request.

bufdev committed:
Move goprotoc logic to app/goprotoc and refactor all control flow to use errors and not os.Exit

bufdev committed:
Ignore capitalized error string check and propagate write errors

jhump commented on app/goprotoc/goprotoc.go:
```diff
@@ -0,0 +1,316 @@
+// Run runs the program and returns the exit code.
+func Run(args []string, stdin io.Reader, stdout io.Writer, stderr io.Writer) int {
+	if err := run(args, stdin, stdout, stderr); err != nil {
+		if message := err.Error(); message != "" {
```
This is a bit strange. Have you encountered a case where code generated a blank error?

bufdev replied:
Yes all the time, it's very annoying when it happens, you'd be surprised - people do do `errors.New("")` or `errors.New(someStringTheyExpectToBeNonEmpty)` a decent amount.

jhump replied:
Strange. I have _never_ seen that. That is unfortunate practice. Though, TBH, it feels like the wrong thing to do to just print nothing but exit with a non-zero code. That would really frustrate me as a user of a command-line tool. You could skip this check, and the result is the same (well, you'd get a single blank line of output instead of nothing -- not materially different to a user).

If you're really concerned about it (FWIW, I wouldn't be), maybe at least do _something_ -- like `fmt.Printf("unexpected error: %#v\n", err)` or panic...

bufdev replied:
I'll update to that - I personally prefer printing nothing, since in theory this is what is intended, but I'll print `"unexpected error"` - in this case, the above will always print `"unexpected error: "`, since we know we have an empty error

bufdev replied:
Done

jhump commented on app/goprotoc/free_fields.go:
```diff
@@ -8,31 +8,43 @@ import (
 
 	unused := computeFreeRanges(md)
-	fmt.Fprintf(w, "%- 35s free:", md.GetFullyQualifiedName())
+	if _, err := fmt.Fprintf(w, "%- 35s free:", md.GetFullyQualifiedName()); err != nil {
```
Ugh. This is the main reason I don't enable the `errcheck` analyzer. This makes code so unreadable, and the cases where process output encounters I/O error is so low that I usually prefer to live with the consequences of silent failure to write output. (99% of the time, if stdout fails here, stderr will fail later, too...)

jhump replied:
This makes me hate the error as return value pattern in Go (and lack of generics, which makes libraries with generic, concise error-handling impossible).

bufdev replied:
It is a low chance, but if I'm testing for 100% correctness, it's worth doing/a one-time annoyance. I know it's stupid but you gotta do it if you really want to be sure.

jhump commented on Makefile:
```diff
@@ -29,10 +29,13 @@ checkgofmt:
 staticcheck:
 	@go get honnef.co/go/tools/cmd/staticcheck
-	staticcheck ./...
+	staticcheck -checks 'all -ST1005' ./...
```
:/

Maybe add a comment to disable this just in the violating file?

jhump replied:
https://staticcheck.io/docs/#ignoring-problems

bufdev replied:
Yea I'll figure out a nicer way to do this that just ignores that one file.

bufdev replied:
Done

jhump reviewed (commented):
This mostly looks good. I know you're making the error handling more robust, but damn does it hurt the readability in some places.

Let's at least tighten up the one staticcheck exception.

bufdev committed:
Update staticcheck to ignore specific errors and get errcheck and golint passing

bufdev commented:
I made a couple upgrades while I'm at it:

- Made it so that `staticcheck` only ignores specific errors per https://staticcheck.io/docs/#ignoring-problems
- With the work on returning `Write` errors, `errcheck ./...` was close to passing, so I made it pass. I also updated the `Makefile` target - the target was `errcheck ./` which actually errored since there are no Golang files in the root of the repository. This updated it to be `errcheck ./...`, which is now called as part of CI.
- Same change to `golint` - `golint` was being called as `golint ./` which did nothing. It was being called during CI already, but wasn't doing anything since it didn't find any files at the root, so I changed it to `golint ./...` and fixed the couple things to make it now pass.

bufdev committed:
Update goprotoc to print "unexpected error" if the returned error is empty

jhump commented on app/goprotoc/goprotoc.go:
```diff
@@ -51,12 +51,15 @@ func run(args []string, stdin io.Reader, stdout io.Writer, stderr io.Writer) err
 	}
 
 	if len(opts.inputDescriptors) > 0 && len(opts.includePaths) > 0 {
+		//lint:ignore ST1005 command return value
```
There are enough of these that perhaps a `lint:file-ignore` at the top would be better.

bufdev replied:
Done

jhump commented on app/goprotoc/goprotoc.go:
```diff
@@ -286,11 +293,11 @@ func saveDescriptor(dest string, fds []*desc.FileDescriptor, includeImports, inc
 	}
-	if b, err := proto.Marshal(&fdSet); err != nil {
+	b, err := proto.Marshal(&fdSet)
+	if err != nil {
```
This is actually a significant reason of why I have `golint` disabled in every project I have. I simply disagree very strongly with how draconian it wants to enforce this particular rule (and a few others). The `errcheck` one is less bad: it can at least be configured to ignore certain methods/patterns, which is nice.

In this particular case, this change is fine. But there are simply cases where typically accepted programming best practices -- to keep a variable's lifetime/scope as short as possible -- are flagged as lint violations. That's just terrible.

FWIW, I think even the authors of golint would say it's an anti-pattern to put it in your CI checks. This is from the official docs: https://github.com/golang/lint#purpose

> In short, this tool is not, and will never be, trustworthy enough for its suggestions to be enforced automatically, for example as part of a build process.

bufdev replied:
I never knew that before - see my below comment. But totally fair.

jhump commented on Makefile:
```diff
@@ -1,5 +1,5 @@
 .PHONY: default
-default: deps checkgofmt vet predeclared staticcheck ineffassign golint test
+default: deps checkgofmt vet predeclared staticcheck ineffassign errcheck golint golint test
```
I'll accept these changes, many of which were made to satisfy the linter. But don't be surprised if I remove `golint` from this list in the future.

bufdev replied:
That's totally fine - I'd argue to keep it until there's something that annoys you, but I get it. I personally like that it only lets me do things in a certain way, but that's not everyone's cup of tea.

jhump reviewed (commented):
One last blocking remark: can you not do a single `file-ignore` directive instead of a dozen `ignore` directives in `goprotoc.go`?

bufdev committed:
Change lint:ignore to lint:file-ignore for staticcheck
